### PR TITLE
feat: `metadata` and `metadataCallback` settings

### PR DESCRIPTION
**Pull request by nd0ut**
Add two settings to set [metadata](https://uploadcare.com/api-refs/rest-api/v0.7.0/#tag/File-Metadata)

## `metadata`
This is plain object with metadata. It should be defined before widget initialization and couldn't be updated after widget initialization.

Could be overrided by `metadataCallback`

Examples:

```js
UPLOADCARE_METADATA = { key: 'value' }
```

```js
new Widget('selector', {
  metadata: { key: 'value' }
})
```

```js
uploadcare.openDialog(null, {
  metadata: { key: 'value' }
})
```

## `metadataCallback` 
This is sync function that called right before upload request. Should return actual metadata object. It allows user to change metadata object during widget work. (this is a real user's use case)

Overrides `metadata` setting, **objects won't be merged, they will be replaced.**

Examples:

```js
UPLOADCARE_METADATA_CALLBACK = () => ({ key: 'value' })
```

```js
new Widget('selector', {
  metadataCallback: () => ({ key: 'value' })
})
```

```js
uploadcare.openDialog(null, {
  metadataCallback: () => ({ key: 'value' })
})
```

## Notes

Since metadata supporting string values only, any non-string value will be converted to string, including booleans, numbers, null and undefined, e.g.

```
null => "null"
undefined => "undefined"
10 => "10"
true => "true"
false => "false"
{ key: 'value' } => "[object Object]"
```

### CODE REVIEW

File: src/files/url.js [hunk 1]
```diff
@@ -49,6 +48,13 @@ class UrlFile extends BaseFile {
       jsonerrors: 1
     }
 
+    const metadata = getMetadataObject(this.settings)
+    if (metadata) {
+      $.each(metadata, (key, value) => {
+        data[`metadata[${key}]`] = String(value)
```
Comment (homm) [inline src/files/url.js:54]: Please ensure no escape required

File: src/files/object.js [hunk 2]
```diff
@@ -113,6 +112,14 @@ class ObjectFile extends BaseFile {
       formData.append('file', this.__file, this.fileName)
       formData.append('file_name', this.fileName)
       formData.append('source', this.sourceInfo.source)
+
+      const metadata = getMetadataObject(this.settings)
+      if (metadata) {
+        $.each(metadata, (key, value) => {
+          formData.append(`metadata[${key}]`, String(value))
+        })
+      }
```
Comment (homm) [inline src/files/object.js:121]: Check if possible to simplify API:

```python
for k, v in getMetadataObject(this.settings):
    data[`metadata[${k}]`] = String(v)

for k, v in getMetadataObject(this.settings):
    formData.append[`metadata[${k}]`] = String(v)
```

**Review by homm — APPROVED** [review]

File: src/files/object.js (hunk 2 above)
Comment (nd0ut) [reply src/files/object.js]: We can't use `for of` and `Object.entries` due to compatibility. So the only way is to use `$.each` to iterate over `k, v`.

I removed `if` condition:

```js
const metadata = getMetadataObject(this.settings)
$.each(metadata, (key, value) => {
  data[`metadata[${key}]`] = String(value)
})
```

It's also possible to reduce lines like this:

```js
$.each(getMetadataObject(this.settings), (key, value) => {
  data[`metadata[${key}]`] = String(value)
})
```

but it looks less clean for me.

Comment (nd0ut) [reply src/files/object.js]: Moved all string casting and iteration logic under the `utils`. 

```js
iterateMetadata(this.settings, (key, value) => {
  data[`metadata[${key}]`] = value
})
```

Comment (homm) [reply src/files/object.js]: I think this is the best option:

```js
$.each(getMetadataObject(this.settings), (key, value) => {
  data[`metadata[${key}]`] = String(value)
})
```

File: src/files/url.js (hunk 1 above)
Comment (nd0ut) [reply src/files/url.js]: Our backend ignores keys containing square brackets. Waiting for Arsenii to resolve this question.

Comment (nd0ut) [reply src/files/url.js]: There is no need to escape or duplicate validation logic. We'll read `Warning` header and print it to the console in the future. In the meantime, we'll make a detailed limitations description in the docs.

File: src/files/object.js (hunk 2 above)
Comment (nd0ut) [reply src/files/object.js]: Yeah, you're right. But I would move type casting to the `getMetadataObject`.

```js
      $.each(getMetadataObject(this.settings), (key, value) => {
        formData.append(`metadata[${key}]`, value)
      })
```

**Pull request merged**